shared/middleware/net_sim_api.rb

```diff
@@ -463,6 +489,17 @@ def delete_messages_for_nodes(shard_id, node_ids)
     message_ids = messages.map {|message| message['id']}
     message_table.delete(message_ids) unless message_ids.empty?
   end
+
+  # Record a custom metric that will be aggregated and sent to New Relic about
+  # once a minute.
+  # @private
+  # @param [String] event_type - unique metric key within NetSimApi
+  # @param [Number] value (default 1) value of measurement, omit if we only care
+  #   about event counts.
+  def record_metric(event_type, value = 1)
+    return unless CDO.newrelic_logging
+    NewRelic::Agent.record_metric("Custom/NetSimApi/#{event_type}", value)
+  end
```

Comment:
Were you able to test new relic logging locally, and if so could you share what steps were necessary to get it working?

Reply:
I haven't tested New Relic logging locally yet.  Blocked on that right at this moment by their site downtime, but once they're back up I think I found [the clue I need here.](https://docs.newrelic.com/docs/agents/ruby-agent/installation/install-new-relic-ruby-agent#Updating_the_Agent)  I think it will involve:

1. Creating a local dashboard/config/newrelic.yml that sets `monitor_mode` true in development and has a valid New Relic key (maybe the development one Paul set up for logging build times).
2. Running the application locally and generating some NetSim traffic.
3. Checking New Relic to see if any metrics were recorded for the development application.

There's also a local audit log option in newrelic.yml that _might_ be an alternative approach to testing - I'll have to experiment a bit.

Reply:
Okay, I've been able to verify some of the metrics being sent via the audit log.  These were my setup steps:

**1. Add `newrelic_logging: true` to my locals.yml**
Otherwise we skip newrelic metrics entirely in our application code.

**2. Create dashboard/config/newrelic.yml**
Making sure to enable the audit log and `monitor_mode` in development. Here's mine:
```yml
common: &default_settings
  license_key: '<redacted>'
  monitor_mode: true
  developer_mode: false
  log_level: info
  audit_log:
    enabled: false
development:
  <<: *default_settings
  app_name: Brad Local Development
  monitor_mode: true
  developer_mode: false
  audit_log:
    enabled: true
    path: /home/brad/newrelic_audit.log
```

**3. Start the local server and generate some traffic**
I used http://localhost-studio.code.org:3000/s/netsim/stage/5/puzzle/1 and made sure to send messages of different sizes, reset the shard a few times, connect and disconnect from various rooms.  Unlike events, metrics get aggregated and sent on intervals by the New Relic agent, so if you want to capture distinct behaviors make sure you separate them by at least 30 seconds or so of downtime.

**4. Check the audit log**
It's very verbose, so grepping for your metrics in particular is highly recommended.  Here are some local results from my change:
```
╰○ egrep -o "Custom/NetSimApi/[^\"]+\",\"scope\":\"\"},\[[^]]+\]" ~/newrelic_audit.log
Custom/NetSimApi/InsertBytes_n","scope":""},[2,198.0,198.0,38.0,160.0,27044.0]
Custom/NetSimApi/InsertRows_n","scope":""},[2,2.0,2.0,1.0,1.0,2.0]
Custom/NetSimApi/InsertBytes_w","scope":""},[1,129.0,129.0,129.0,129.0,16641.0]
Custom/NetSimApi/InsertRows_w","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[2,338.0,338.0,168.0,170.0,57124.0]
Custom/NetSimApi/InsertRows_m","scope":""},[2,2.0,2.0,1.0,1.0,2.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[1,159.0,159.0,159.0,159.0,25281.0]
Custom/NetSimApi/InsertRows_l","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[6,1012.0,1012.0,143.0,194.0,173098.0]
Custom/NetSimApi/InsertRows_m","scope":""},[6,6.0,6.0,1.0,1.0,6.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[3,476.0,476.0,134.0,183.0,76726.0]
Custom/NetSimApi/InsertRows_l","scope":""},[3,3.0,3.0,1.0,1.0,3.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[28,23940.0,23940.0,164.0,971.0,22690528.0]
Custom/NetSimApi/InsertRows_m","scope":""},[28,28.0,28.0,1.0,1.0,28.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[14,11830.0,11830.0,155.0,960.0,11107250.0]
Custom/NetSimApi/InsertRows_l","scope":""},[14,14.0,14.0,1.0,1.0,14.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[64,8686.0,8686.0,126.0,137.0,1179074.0]
Custom/NetSimApi/InsertRows_m","scope":""},[64,64.0,64.0,1.0,1.0,64.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[32,4023.0,4023.0,117.0,126.0,505845.0]
Custom/NetSimApi/InsertRows_l","scope":""},[32,32.0,32.0,1.0,1.0,32.0]
Custom/NetSimApi/InsertBytes_n","scope":""},[2,194.0,194.0,38.0,156.0,25780.0]
Custom/NetSimApi/InsertRows_n","scope":""},[2,2.0,2.0,1.0,1.0,2.0]
Custom/NetSimApi/InsertBytes_w","scope":""},[1,136.0,136.0,136.0,136.0,18496.0]
Custom/NetSimApi/InsertRows_w","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
Custom/NetSimApi/UpdateBytes_n","scope":""},[12,1828.0,1828.0,149.0,157.0,278560.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[1,139.0,139.0,139.0,139.0,19321.0]
Custom/NetSimApi/InsertRows_m","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[1,130.0,130.0,130.0,130.0,16900.0]
Custom/NetSimApi/InsertRows_l","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[9,1239.0,1239.0,135.0,140.0,170605.0]
Custom/NetSimApi/InsertRows_m","scope":""},[9,9.0,9.0,1.0,1.0,9.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[5,642.0,642.0,126.0,130.0,82452.0]
Custom/NetSimApi/InsertRows_l","scope":""},[5,5.0,5.0,1.0,1.0,5.0]
Custom/NetSimApi/UpdateBytes_n","scope":""},[2,305.0,305.0,151.0,154.0,46517.0]
Custom/NetSimApi/InsertBytes_n","scope":""},[12,1170.0,1170.0,35.0,160.0,160950.0]
Custom/NetSimApi/InsertRows_n","scope":""},[12,12.0,12.0,1.0,1.0,12.0]
Custom/NetSimApi/ResetShard","scope":""},[4,4.0,4.0,1.0,1.0,4.0]
Custom/NetSimApi/InsertBytes_w","scope":""},[1,127.0,127.0,127.0,127.0,16129.0]
Custom/NetSimApi/InsertRows_w","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[2,330.0,330.0,164.0,166.0,54452.0]
Custom/NetSimApi/InsertRows_m","scope":""},[2,2.0,2.0,1.0,1.0,2.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[1,152.0,152.0,152.0,152.0,23104.0]
Custom/NetSimApi/InsertRows_l","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
Custom/NetSimApi/InsertBytes_m","scope":""},[16,14748.0,14748.0,200.0,1499.0,18327268.0]
Custom/NetSimApi/InsertRows_m","scope":""},[16,16.0,16.0,1.0,1.0,16.0]
Custom/NetSimApi/InsertBytes_l","scope":""},[8,7270.0,7270.0,188.0,1485.0,8973254.0]
Custom/NetSimApi/InsertRows_l","scope":""},[8,8.0,8.0,1.0,1.0,8.0]
Custom/NetSimApi/ResetShard","scope":""},[1,1.0,1.0,1.0,1.0,1.0]
```
I'm not sure of the exact meanings of the array of values, but they appear to be: [count, total, total, min, max, sum_of_squares].  The values match the ranges I'd expect for various events.

**5. Check New Relic's Dashboard**
My local machine is definitely reporting to New Relic - it shows up as an application in the APM tool:
![image](https://user-images.githubusercontent.com/1615761/31254766-dbed8dc8-a9de-11e7-8a02-8c6a5693e581.png)

However, I haven't been able to find any of my new metrics in the metrics explorer.  I'm wondering if there's a delay for that data being populated or if their recent site issues are affecting anything, or if there's just another configuration setting I've missed.  In any case, the audit log makes me confident that we'll get these metrics all the way through the system once this change is in prod.